indexer: move deduplication functionality purely to the kvindexer

cmwaters opened the pull request:
Closes: https://github.com/tendermint/tendermint/issues/9460

The deduplication logic which prevents older failed transactions from overwriting newer successful transactions called the `Get` method on the `TxIndexer`. For postgresql, this wasn't implemented and would always return an error. Errors were only logged which meant that the postgresql would batch `nil` for that height` and then continue. This resolves the issue by moving the logic purely within the kvindexer

jmalicevic commented on state/txindex/kv/kv.go:
```diff
@@ -102,12 +102,30 @@ func (txi *TxIndex) AddBatch(b *txindex.Batch) error {
+//
+// If a transaction is indexed with the same hash as a previous transaction, it will
+// be overwritten unless the tx result was NOT OK and the prior result was OK i.e.
+// more transactions that successfully executed overwrite transaction that failed
```
```suggestion
// more transactions that successfully executed overwrite transactions that failed
```

jmalicevic commented on state/txindex/kv/kv.go:
```diff
@@ -102,12 +102,30 @@ func (txi *TxIndex) AddBatch(b *txindex.Batch) error {
+// If a transaction is indexed with the same hash as a previous transaction, it will
+// be overwritten unless the tx result was NOT OK and the prior result was OK i.e.
+// more transactions that successfully executed overwrite transaction that failed
+// or successfult yet older transactions.
```
```suggestion
// or successful yet older transactions.
```

cmwaters commented:
> Could you please add a test here to test specifically for the problem outlined in https://github.com/tendermint/tendermint/issues/9460?

I've added a test. As part of it, I've added a flag in the constructor which will terminate the indexing service upon the first error. That way I can detect that the postgres indexer has successfully processes the transactions

samricotta commented on state/txindex/indexer_service.go:
```diff
@@ -74,24 +75,38 @@ func (is *IndexerService) OnStart() error {
-			batch.Ops, err = DeduplicateBatch(batch.Ops, is.txIdxr)
-			if err != nil {
-				is.Logger.Error("deduplicate batch", "height", height)
+				is.Logger.Info("indexed block exents", "height", height)
```
Why is the error no longer needed here. Is this only because we are getting rid of the "deduplicate batch"

cmwaters commented:
I will add the changelog for the v0.34 and v0.37 backports

cmwaters commented on state/indexer/sink/psql/psql_test.go:
```diff
@@ -196,6 +197,55 @@ func TestIndexing(t *testing.T) {
+			Height: 1,
+			Index:  uint32(1),
+			Tx:     types.Tx("bar"),
+			Result: abci.ResponseDeliverTx{Code: 1},
```
Yes, previously it would error

cmwaters commented on state/indexer/sink/psql/psql_test.go:
```diff
@@ -196,6 +197,55 @@ func TestIndexing(t *testing.T) {
+		err = eventBus.PublishEventTx(types.EventDataTx{TxResult: *txResult2})
+		require.NoError(t, err)
+
+		time.Sleep(100 * time.Millisecond)
```
Yeah, currently the indexer is such an opaque system to debug. One improvement I thought of would be to track the latest height indexed, then we could call to make sure that it had indexed these heights without error

pratikbin commented:
I couldn't find the docs for creating initial migration or relation for index in DB
